app/src/main/java/com/willowtree/vocable/presets/adapter/AddToCategoryPickerViewModel.kt

```diff
@@ -0,0 +1,154 @@
+
+    private val liveCategoryMap = MutableLiveData<Map<Category, Boolean>>()
+    val categoryMap: LiveData<Map<Category, Boolean>> = liveCategoryMap
+
+    private val liveShowPhraseAdded = MutableLiveData<Boolean>()
+    val showPhraseAdded: LiveData<Boolean> = liveShowPhraseAdded
+
+    private val liveShowPhraseDeleted = MutableLiveData<Boolean>()
+    val showPhraseDeleted: LiveData<Boolean> = liveShowPhraseDeleted
+
+    private var userCategories = listOf<Category>()
+
+    // a list that maps categories to the phrase.  The phrase will not change, so we just need
+    // to map the category id to the phrase object that was added to it, so we can delete it later
+    private var savedPhrases = mutableMapOf<String, Phrase>()
+
+    fun handleCategoryToggled(phraseString: String, category: Category, isChecked: Boolean) {
+        backgroundScope.launch {
+            // if the user checked the switch
+            if (isChecked) {
+                // add it to the category
+                addNewPhrase(phraseString, category)
+            } else { // if they un-checked the switch
+                // delete it from the category
+                deletePhraseFromCategory(phraseString, category)
+            }
+        }
+    }
+
+    fun buildCategoryList(phraseString: String) {
+        // map category to whether or not the phrase is in that category
+        val phraseMap = mutableMapOf<Category, Boolean>()
+
+        backgroundScope.launch {
+            userCategories = presetsRepository.getUserGeneratedCategories()
+
+            // for each category in the list
+            userCategories.forEach { category ->
+                // get all the phrases associated with it
+                val phrases = presetsRepository.getPhrasesForCategory(category.categoryId)
+
+                // for each phrase in the category
+                phrases.forEach {
+                    // if we find a match
+                    if (localizedResourceUtility.getTextFromPhrase(it) == phraseString) {
+                        // add it to the map
+                        phraseMap[category] = true
+                    }
+                }
+
+                // if we get through the loop without adding anything, the phrase isn't in there
+                if (phraseMap[category] == null) {
+                    phraseMap[category] = false
+                }
+            }
+
+            liveCategoryMap.postValue(phraseMap)
+        }
+    }
+
+    private fun deletePhraseFromCategory(phraseString: String, category: Category) {
+        backgroundScope.launch {
+            var phrase: Phrase? = null
+
+            val phrasesInCategory = presetsRepository.getPhrasesForCategory(category.categoryId)
+            for (p in phrasesInCategory) {
+                if (localizedResourceUtility.getTextFromPhrase(p) == phraseString) {
+                    phrase = p
+                }
+            }
```

Comment:
You could probably just use a `findLast {  }` function to set `phrase` here.

```
val phrasesInCategory = presetsRepository.getPhrasesForCategory(category.categoryId)
val phrase = phrasesInCategory.findLast { localizedResourceUtility.getTextFromPhrase(p) == phraseString }
```

My bad if that's not what you're trying to attempt here and I think it might return a nullable as well? Maybe there's a `findLastOrNull {  }`